File: dailyQuestion/2021/2021-02/02-27/golang/recursion.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	s := "ababbc"
	res := longestSubstring(s, 2)

	fmt.Println(res)
}

func longestSubstring(s string, k int) (res int) {
	if s == "" {
		return
	}
	cnt := [26]int{}
	for _, ch := range s {
		cnt[ch-'a']++
	}

	var split byte
	for i, c := range cnt {
		if 0 < c && c < k {
			split = byte(i) + 'a'
			break
		}
	}
	if split == 0 {
		return len(s)
	}
	for _, subStr := range strings.Split(s, string(split)) {
		res = max(res, longestSubstring(subStr, k))
	}
	return
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: dailyQuestion/2021/2021-02/02-27/python/recursion.py
class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        res = 0
        if not s:
            return res
        cnt = [0] * 26
        for item in s:
            cnt[ord(item) - 97] += 1
        split = ""
        for index, item in enumerate(cnt):
            if item > 0 and item < k:
                split = chr(index + 97)
                break
        if split == "":
            return len(s)
        for item in s.split(split):
            res = max(res, self.longestSubstring(item, k))
        return res


if __name__ == "__main__":
    s = "ababbc"
    so = Solution()
    res = so.longestSubstring(s, 2)
    print(res)
